### ISSUE TITLE
Implementation question - in-device sliding window?

### ISSUE THREAD

**Issue by Humancell** [mrmp17/PLab-Vibration-Probe-FW#1]
This is a very cool device!  I'm curious how much local storage might be in the device for this possible application:

We want to monitor for large events (e.g. local blasting at a mine, or even an earthquake) and so we want the device to be locally sampling at a high rate, and storing a "window" of samples that could be a few seconds (5-10 max?).  Upon a threshold level being hit, it would then send the "window" surrounding that event, half the window size before the event and half after the event.

Our goal is to have this processing occur within the device, and not to have to implement this in an external MCU.

I completely understand that RAM within the device might be limited, but wanted to ask for your thoughts on this.  Our goal is to be able to report "snippets" of event data that surround events that peak above a threshold.  Pushing more intelligence into the device itself.

Thoughts?

**Comment by nplan** [comment]
The device has enough storage for around 4k samples (8k without gyro data) so it might be possible depending on the desired sampling frequency.

But based on your example application, I see no reason why on-device monitoring and storage would be better than doing it on the computer. The device has to be connected to a computer either way, so I think it makes more sense to keep the on-device firmware as simple as possible and do the event monitoring, processing and storage on the computer.

In other words:  as a user, you always interact with the device through the Python API. So from your perspective it makes no difference if the monitoring is happening on the device itself or on the computer.

Or is there another reason for this idea that I'm not seeing?

**Comment by Humancell** [comment]
Yes, I forgot to mention, I was not planning on having any "computer" present.  I'm considering simply attaching it to a microcontroller, or a cellular module, where I'd be running some basic communications code.

So when an event was detected, and the window of data was tagged, then the microcontroller would fetch the window from the device and upload the window of data to the cloud.

This would be a very lightweight device, maybe running on solar.

**Comment by nplan** [comment]
Currently, the only supported host is a computer running Windows / Linux / macOS. For your use case, a Raspberry Pi would be a great choice.

**Comment by Humancell** [comment]
One more question ... have you documented the protocol that is going over the host connection?

**Comment by nplan** [comment]
No, it's not documented. It's protobuf over a serial port.

The latest version 1.1.0 supports triggering on events based on RMS. But all the computation is still done on the host.
